design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+    The cluster health will be verified as a whole before moving to the next monitor.
+1. **OSDs:** The OSD pods will be upgraded in a rolling fashion after the monitors.  **For each** OSD, the following actions will take place:
+    1. The `image` field of the pod template spec will be updated to the new version number.
+    1. The lifecycle management of OSDs can be done either as a whole by a single daemon set or individually by a replica set per OSD.
+    In either case, each individual OSD pod will be terminated so that its managing controller will respawn a new pod on the new version in its place.
+    1. The controller will verify that each OSD is running the new version and that they return to the `UP` and `IN` statuses.
+    Placement group health will also be verified to ensure all PGs return to the `active+clean` status before moving on.
+1. **Ceph Manager:** The Ceph manager pod will be upgraded next by updating the `image` field on the pod template spec.
+The deployment that is managing the pod will then terminate it and start a new pod running the new version.
+    1. The upgrade controller will verify that the new pod is on the new version, in the `Running` state and that the manager instance shows as `Active` in the Ceph status output.
+1. If the user has installed optional components, such as object storage (**RGW**) or shared file system (**MDS**), they will also be upgraded to the new version.
+They are both managed by deployments, so the upgrade controller will update the `image` field in their pod template specs which then causes the deployment to terminate old pods and start up new pods on the new versions to replace them.
+    1.  Cluster health and object/file functionality will be verified before the upgrade controller moves on to the next instances.
+
+### **Upgrade Health Verification**
+As mentioned previously, the manual health verification steps found in the [upgrade user guide](../Documentation/upgrade.md#health-verification) will be used by the upgrade controller, in an automated fashion, to ensure the cluster is healthy before proceeding with the upgrade process.
+This approach of upgrading one component, verifying health and stability, then upgrading the next component can be viewed as a form of [canary deployment](https://kubernetes.io/docs/concepts/cluster-administration/manage-deployment/#canary-deployments).
+
+Here is a quick summary of the standard health checks the upgrade controller should perform:
+* All pods are in the `Running` state and have few, if any, restarts
+  * No pods enter a crash loop backoff
+* Overall status: The overall cluster status is `OK` and there are no warning or error status messages displayed.
+* Monitors:  All of the monitors are `in quorum` and have individual status of `OK`.
+* OSDs: All OSDs are `UP` and `IN`.
+* MGRs: All Ceph managers are in the `Active` state.
+* Placement groups: All PGs are in the `active+clean` state.
```

Comment:
should we also perform a sanity test by creating a PVC and attaching it to a dummy pod? This will ensure that the attach/detach is working and will test the provisioner and rook-agaent.

Reply:
creating dummy pods and test PVCs is awkward. Health checks seem sufficient, as long as we find a complete set of health checks.